common/txmgr/address_state.go

```diff
@@ -0,0 +1,250 @@
+}
+
+// newAddressState returns a new addressState instance with initialized transaction state
+func newAddressState[
+	CHAIN_ID types.ID,
+	ADDR, TX_HASH, BLOCK_HASH types.Hashable,
+	R txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH],
+	SEQ types.Sequence,
+	FEE feetypes.Fee,
+](
+	lggr logger.SugaredLogger,
+	chainID CHAIN_ID,
+	fromAddress ADDR,
+	maxUnstarted uint64,
+	txs []txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE],
+) *addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE] {
+	if maxUnstarted == 0 {
+		panic("new_address_state: MaxUnstarted queue size must be greater than 0")
+	}
+
+	// Count the number of transactions in each state to reduce the number of map resizes
+	counts := map[txmgrtypes.TxState]int{
+		TxUnstarted:               0,
+		TxInProgress:              0,
+		TxUnconfirmed:             0,
+		TxConfirmedMissingReceipt: 0,
+		TxConfirmed:               0,
+		TxFatalError:              0,
+	}
+	var idempotencyKeysCount int
+	var txAttemptCount int
+	for _, tx := range txs {
+		counts[tx.State]++
+		if tx.IdempotencyKey != nil {
+			idempotencyKeysCount++
+		}
+		if len(tx.TxAttempts) > 0 {
+			txAttemptCount += len(tx.TxAttempts)
+		}
+	}
+
+	as := addressState[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]{
+		lggr:        lggr,
+		chainID:     chainID,
+		fromAddress: fromAddress,
+
+		idempotencyKeyToTx:         make(map[string]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], idempotencyKeysCount),
+		unstartedTxs:               queues.NewTxPriorityQueue[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE](int(maxUnstarted)),
+		inprogressTx:               nil,
+		unconfirmedTxs:             make(map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], counts[TxUnconfirmed]),
+		confirmedMissingReceiptTxs: make(map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], counts[TxConfirmedMissingReceipt]),
+		confirmedTxs:               make(map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], counts[TxConfirmed]),
+		allTxs:                     make(map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], len(txs)),
+		fatalErroredTxs:            make(map[int64]*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], counts[TxFatalError]),
+		attemptHashToTxAttempt:     make(map[TX_HASH]*txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], txAttemptCount),
+	}
+
+	// Load all transactions supplied
+	for i := 0; i < len(txs); i++ {
+		tx := txs[i]
+		switch tx.State {
+		case TxUnstarted:
+			as.unstartedTxs.AddTx(&tx)
+		case TxInProgress:
+			as.inprogressTx = &tx
+		case TxUnconfirmed:
+			as.unconfirmedTxs[tx.ID] = &tx
+		case TxConfirmedMissingReceipt:
+			as.confirmedMissingReceiptTxs[tx.ID] = &tx
+		case TxConfirmed:
+			as.confirmedTxs[tx.ID] = &tx
+		case TxFatalError:
+			as.fatalErroredTxs[tx.ID] = &tx
+		default:
+			panic("unknown transaction state")
```

Comment:
@prashantkumar1982 @poopoothegorilla does panic make sense here? I doubt we will ever hit this, but if by some miracle we do, wouldn't it be better to just drop the TX, log it, and proceed? Or is this error only possible through some sort of dev error that we want to catch ASAP?